students/jbutts/Lesson05/mailroom3.py

```diff
@@ -0,0 +1,199 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+import tempfile
+
+'''
+revised mailroom app using dict switch and dict as a data structure (we used a list 
+in the first version.
+
+Added ability to create letters as text files, in a temp directory, in batch.
+
+Added sorting abilities here. I originally missed this in the first mailroom assignment.
+
+'''
+
+
+def populate_data(donor_db):
+    '''
+    Populate dictionary with initial data
+    '''
+
+    donor_db = add_contribution("Daenerys Targaryen", [100.00, 1000], donor_db)
+    donor_db = add_contribution("Arya Stark", [65.00, 45.92, 1004, 2008, 7777], donor_db)
+    donor_db = add_contribution("Melisandre", [19.00, 100000, 100000, 100000], donor_db)
+    donor_db = add_contribution("Cersei Lannister", [10.00], donor_db)
+    donor_db = add_contribution("Daenerys Targaryen", [400, 500, 700, 5555], donor_db)
+    donor_db = add_contribution("Bran Stark", [666], donor_db)
+    donor_db = add_contribution("Bran Stark", 777, donor_db) # We even handle for floats instead
+    # of lists, just in case
+    return donor_db
+
+
+def show_donors(donor_db):
+    '''
+    Write a report to screen of donors and information about how they've donated
+
+    Return values in descending order from the sum of their historical contributions
+    '''
+
+    # Write header row
+    row = "Donor", "Total", "Gifts", "Average"
+    print("{:<25s}{:<20s}{:<10s}{:<13s}".format(*row))
+    # Write corresponding ===='s
+    print(''.join(("=" * 24 + " ",  # donor
+                   "=" * 19 + " ",  # total
+                   "=" * 9 + " ",   # gifts
+                   "=" * 12)))      # average
+
+    #  Get a sorted list of donors names sorted by their contributions, descending
+    ordered_donors = sorted(donor_db, key=lambda k: sum(donor_db[k]), reverse=True)
+    for donor in ordered_donors:
+        #  Iterate through the dict to match the sorted list.
+        for key, value in donor_db.items():
+            if key == donor:
+                # alphabetical by keys. Write out rows: "Donor", "Total", "Gifts", "Average"
+                row = key, sum(value), len(value), sum(value)/len(value)
+                print("{:<25s}${:<19.2f}{:<10d}${:<13.2f}".format(*row))
+
+
+
+def list_donors(donor_db):
+    '''
+    Write a numbered list for the purpose of selecting a donor
+
+    Use a list comprehension to print the numbered list
+    '''
+    # List current donors in the dictionary as '#. First Last'
+    print("{:<4} {:<7}".format("No.", "Name"))
+    print("{} {}".format("=" * 4, "=" * 7))
+    [print("{:<4} {:<7}". format(i + 1, donor)) for i, donor in enumerate(sorted(donor_db))]
```

Comment:
this is incorrect usage of comprehension - you are generating an empty list, in a comprehension you should be generating list (dict,set/generator) and using the returned values.